### PR TITLE
feat(cargo): support sparse registry indices

### PR DESCRIPTION
**Pull request by tofay**
## Changes

Allow Renovate to update dependencies from private sparse Rust crate registries.

## Context

Sparse registries, documented at https://doc.rust-lang.org/nightly/cargo/reference/registry-index.html#sparse-protocol, use a HTTP file based index rather than git. Per https://blog.rust-lang.org/inside-rust/2023/01/30/cargo-sparse-protocol.html they will stabilize in the next Rust release (1.68). I use them at work, and have started using renovate to keep Rust dependencies up to date.

The directory structure of sparse registries (in particular the index suffixing e.g `fo/ob/foobar) is identical to existing registries. Given that renovate already uses HTTP file based access to the crates.io index on github, this PR mainly consists of detecting the `sparse+` prefix of index URLs and reusing the crates-io code path.

(I've not touched on authentication here as sparse registries aren't necessarily authenticated. I am running renovate self hosted and have this code working against authenticated sparse registries by:

- configuring a host URL for renovate's HTTP calls to the index
- configuring cargo to use authentication via environment variable
)

I found https://github.com/renovatebot/renovate/pull/18941 when looking for related issues, though the changes in that draft PR looked specific to JFrog artifactory.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

I've been running this self-hosted against a few dozen private Azure DevOps repositories that use dependencies from a sparse registry for a couple of weeks, and had PRs raised succesfully.

### CODE REVIEW

File: docs/usage/rust.md
Comment (HonkingGoose) [inline docs/usage/rust.md:34]:
```suggestion
Since sparse registries are HTTP based, authentication for Renovate can be configured via additional `hostRules`.
```

File: docs/usage/rust.md
Comment (HonkingGoose) [inline docs/usage/rust.md:33]:
```suggestion
Renovate can update dependencies hosted on a private sparse crate registry, read [Rust's Cargo book](https://doc.rust-lang.org/beta/cargo/reference/registry-index.html#sparse-protocol) to learn more.
```

Changes:

- Fix link syntax
- Descriptive name for the link

File: docs/usage/rust.md
Comment (HonkingGoose) [inline docs/usage/rust.md:36]:
```suggestion
Before Renovate can update Cargo lockfiles for projects using dependencies from sparse registries you must:

- Use Rust `1.68` or later
- Enable the `sparse-registry` feature in your local Rust toolchain
```

How about something like this?

The current sentence is too long, and has too much information in one go. So we should split things up into smaller parts. 😉

File: lib/modules/datasource/crate/types.ts
Comment (HonkingGoose) [inline lib/modules/datasource/crate/types.ts:22]:
```suggestion
  /** Download location of crates, as specified in the registry's config.json */
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
@Turbo87 could you please review?

File: docs/usage/rust.md
Comment (HonkingGoose) [reply docs/usage/rust.md:36]: > Perhaps omit all mention of the sparse-registry feature, which is actually a compilation flag so unlikely something most users can control?

It sounds we have two groups of people here:

1. General public who uses Renovate to update Rust dependencies
1. Maintainers of Rust repositories who want Renovate to work properly for their published packages

Sounds like the compilation flag `sparse-registry` is aimed at number 2 then?

I'll let the Renovate maintainers decide what to keep in the docs and what to drop. 😉 

> The only oddity with that is that Rust 1.68 isn't out yet, but anyone using sparse registries before then is likely using a beta/nightly toolchain, or a self-built with sparse-registry built in (I fall into this camp :)), and can figure out how to get it working.

I'm okay with mentioning the Rust version `1.68` ahead of time. Are you sure that the `sparse-registry` feature will be included in version `1.68`?

Comment (tofay) [reply docs/usage/rust.md:36]: > Are you sure that the sparse-registry feature will be included in version 1.68?

Yes, at least that's still the current plan: https://blog.rust-lang.org/inside-rust/2023/01/30/cargo-sparse-protocol.html

File: lib/modules/datasource/crate/index.spec.ts
Comment (viceice) [inline lib/modules/datasource/crate/index.spec.ts:272]: no snapshots, read our dev best practices guild inside docs folder of this repo.

File: lib/modules/datasource/crate/index.ts
Comment (viceice) [inline lib/modules/datasource/crate/index.ts:29]: why another http? can't you simply use the existing from base class?
```suggestion
```

File: lib/modules/datasource/crate/index.ts
Comment (viceice) [inline lib/modules/datasource/crate/index.ts:10]:
```suggestion
```
see below

File: lib/modules/datasource/crate/types.ts
Comment (viceice) [inline lib/modules/datasource/crate/types.ts:23]: please use full name here for readability

**Review by viceice — CHANGES_REQUESTED** [review]

**Comment by viceice** [comment]
related discussion
- https://github.com/renovatebot/renovate/discussions/18872

**Review by Turbo87 — COMMENTED** [review]
looks alright to me on first glance. I haven't tried it out locally though and my experience is limited to the sparse implementation for crates.io itself.

**Comment by tofay** [comment]
closing as https://github.com/renovatebot/renovate/pull/21187 has landed

**Pull request closed without merging**
